File: lab1_2.kt
import java.util.Arrays

class Sorter {

    var intArray = arrayOf(40, 8, 11, 2, 9)

    fun swap(arr: IntArray, a: Int, b: Int) {
        val temp = arr[a]
        arr[a] = arr[b]
        arr[b] = temp
    }

    fun partition(arr: IntArray, low: Int, high: Int): Int {
        val pivot = arr[high]
        var i = low - 1

        for (j in low until high) {
            if (arr[j] < pivot) {
                i++
                swap(arr, i, j)
            }
        }

        swap(arr, i + 1, high)
        return i + 1
    }

    fun quickSort(arr: IntArray, low: Int, high: Int) {
        if (low < high) {
            val pi = partition(arr, low, high)

            quickSort(arr, low, pi - 1)
            quickSort(arr, pi + 1, high)
        }
    }

}

fun main() {
    val arr = intArrayOf(40, 8, 11, 2, 9)
    val sort = Sorter()

    sort.quickSort(arr,0, arr.size - 1)

    println(Arrays.toString(arr))
}

File: labs/lab1/lab1_2.cpp
#include <stdio.h>
#include <iostream>
using namespace std;

class Sorter
{
public:
    int low = 0;

    void swap(int a, int b)
    {
        int temp;

        temp = a;
        a = b;
        b = temp;
    }

    int separe(int v[], int low, int high)
    {
        int pivot;
        pivot = v[high];

        int i, j;
        i = low - 1;

        for (j = low; j <= high - 1; j++)
        {
            if (v[j] < pivot)
            {
                i++;
                swap(v[i], v[j]);
            }
        }

        swap(v[i + 1], v[high]);

        return (i + 1);
    }

    void quickSort(int v[], int low, int high)
    {
        int sepIndex;

        if (low < high)
        {
            sepIndex = separe(v, low, high);

            quickSort(v, low, sepIndex - 1);

            quickSort(v, sepIndex + 1, high);
        }
    }

    void printArray(int arr[], int n)
    {
        int i;
        for (i = 0; i < n; i++)
            cout << arr[i] << " ";
        cout << endl;
    }
};

int main()
{
    int arr[5] = {40, 8, 11, 2, 99};
    int n = 5;

    Sorter sorting;

    sorting.quickSort(arr, 0, n);
    sorting.printArray(arr, n);

    return 0;
}